src/lib.rs

```diff
@@ -1,40 +1,97 @@
-extern crate byteorder;
-#[macro_use]
-extern crate enum_primitive;
-extern crate num;
+use std::any::type_name;
+use std::io::{Read, Seek};
 
-use std::fmt;
-use std::io::Result;
+use num_traits::FromPrimitive;
+use tiff::decoder::{Decoder, DecodingResult};
+use tiff::tags::Tag;
+use tiff::TiffResult;
 
-pub mod geotiff;
-mod lowlevel;
-mod reader;
+use crate::image_data::*;
 
-pub use geotiff::TIFF;
-use reader::*;
+mod image_data;
 
-/// The GeoTIFF library reads `.tiff` files.
+macro_rules! assert_type {
+    ($result: expr, $actual: ty, $expected: ty) => {
+        $result
+            .ok_or_else(|| {
+                format!(
+                    "Cannot represent {} as {}",
+                    type_name::<$actual>(),
+                    type_name::<$expected>()
+                )
+            })
+            .unwrap()
+    };
+}
+
+/// The basic GeoTIFF struct. This includes any metadata as well as the actual image data.
 ///
-/// It is primarily used within a routing application that needs to parse digital elevation models.
-/// As such, other use cases are NOT tested (for now).
-impl TIFF {
-    /// Opens a `.tiff` file at the location indicated by `filename`.
-    pub fn open(filename: &str) -> Result<Box<TIFF>> {
-        let tiff_reader = TIFFReader;
-        tiff_reader.load(filename)
-    }
+/// The image data has a size of width * length * num_samples
+#[derive(Debug)]
+pub struct GeoTIFF {
+    pub dimensions: (usize, usize),
+    pub num_samples: usize,
```

Comment:
I know you've gone back and forth on this, but just want to check if we're ok with settling on `num_samples` and not `num_bands`?

Reply:
Yeah, I wasn't sure what is the most appropriate terminology to use here
- On the `TIFF` file format level, they refer to `samples`.
- In their description of the raster data model, [GDAL](https://gdal.org/en/latest/user/raster_data_model.html) calls it `bands`.
- Image manipulation software usually calls it `channels.`
- `layer` is another possible term mentioned by GDAL.

The reason why I chose `sample` over `band` was because the origin of `band` comes from the bandwidth of the color spectrum that the sample is supposed to cover, which does not apply to all scenarios where raster datasets are used. I have no absolute preference here; we can also change the name later if we discover that it does not fit.

Reply:
Yeah, I'm torn as well, the remote sensing part of me wants to call it `bands`, but the computer vision part of me likes `channels`. Let's stick with `samples` then following TIFF, though it feels very statistical for some reason :laughing: